- Description: dasd: DIAG access does not work when compiling DIAG as module.
  Symptom:     On systems where the DASD DIAG component was compiled as a kernel
               module, trying to set a DASD device online after activating
               use_diag (SysFS attribute) will always fail.
  Problem:     The DASD-internal check for availability of the DIAG component is
               incorrect in case DIAG was compiled as module.
  Solution:    Fix the DIAG check.
  Problem-ID:  -
  Archs:       s390-31

--- linux-2.5/drivers/s390/block/dasd.c	10 May 2004 12:46:18 -0000	1.137.2.4
+++ linux-2.5/drivers/s390/block/dasd.c	25 May 2004 13:37:59 -0000	1.137.2.5
@@ -7,7 +7,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999-2001
  *
- * $Revision: 1.137.2.4 $
+ * $Revision: 1.137.2.5 $
  */
 
 #include <linux/config.h>
@@ -37,6 +37,7 @@
  * SECTION: exported variables of dasd.c
  */
 debug_info_t *dasd_debug_area;
+struct dasd_discipline *dasd_diag_discipline_pointer;
 
 MODULE_AUTHOR("Holger Smolinski <Holger.Smolinski@de.ibm.com>");
 MODULE_DESCRIPTION("Linux on S/390 DASD device driver,"
@@ -1990,6 +1991,8 @@
 
 	DBF_EVENT(DBF_EMERG, "%s", "debug area created");
 
+	dasd_diag_discipline_pointer = NULL;
+
 	rc = devfs_mk_dir("dasd");
 	if (rc)
 		goto failed;
@@ -2022,6 +2025,7 @@
 module_exit(dasd_exit);
 
 EXPORT_SYMBOL(dasd_debug_area);
+EXPORT_SYMBOL(dasd_diag_discipline_pointer);
 
 EXPORT_SYMBOL(dasd_add_request_head);
 EXPORT_SYMBOL(dasd_add_request_tail);

--- linux-2.5/drivers/s390/block/dasd_int.h	20 Apr 2004 13:28:03 -0000	1.56.2.1
+++ linux-2.5/drivers/s390/block/dasd_int.h	25 May 2004 13:38:00 -0000	1.56.2.2
@@ -6,7 +6,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999,2000
  *
- * $Revision: 1.56.2.1 $
+ * $Revision: 1.56.2.2 $
  */
 
 #ifndef DASD_INT_H
@@ -260,12 +260,7 @@
 	int (*fill_info) (struct dasd_device *, struct dasd_information2_t *);
 };
 
-extern struct dasd_discipline dasd_diag_discipline;
-#ifdef CONFIG_DASD_DIAG
-#define dasd_diag_discipline_pointer (&dasd_diag_discipline)
-#else
-#define dasd_diag_discipline_pointer (0)
-#endif
+extern struct dasd_discipline *dasd_diag_discipline_pointer;
 
 struct dasd_device {
 	/* Block device stuff. */

--- linux-2.5/drivers/s390/block/dasd_diag.c	4 Mar 2004 13:06:21 -0000	1.34
+++ linux-2.5/drivers/s390/block/dasd_diag.c	25 May 2004 13:38:00 -0000	1.34.2.1
@@ -6,7 +6,7 @@
  * Bugreports.to..: <Linux390@de.ibm.com>
  * (C) IBM Corporation, IBM Deutschland Entwicklung GmbH, 1999,2000
  *
- * $Revision: 1.34 $
+ * $Revision: 1.34.2.1 $
  */
 
 #include <linux/config.h>
@@ -35,6 +35,8 @@
 
 MODULE_LICENSE("GPL");
 
+struct dasd_discipline dasd_diag_discipline;
+
 struct dasd_diag_private {
 	struct dasd_diag_characteristics rdc_data;
 	struct dasd_diag_rw_io iob;
@@ -489,6 +491,7 @@
 
 	ctl_set_bit(0, 9);
 	register_external_interrupt(0x2603, dasd_ext_handler);
+	dasd_diag_discipline_pointer = &dasd_diag_discipline;
 	return 0;
 }
 
@@ -503,6 +506,7 @@
 	}
 	unregister_external_interrupt(0x2603, dasd_ext_handler);
 	ctl_clear_bit(0, 9);
+	dasd_diag_discipline_pointer = NULL;
 }
 
 module_init(dasd_diag_init);

